kernel/src/scan/mod.rs

```diff
@@ -218,20 +214,30 @@ impl Scan {
                 let read_result = if partition_fields.is_empty() {
                     read_result
                 } else {
-                    let mut fields = Vec::with_capacity(partition_fields.len() + len);
+                    debug!("Adding back in partition columns");
```

Comment:
Quick note to explain what had to change here. We previously just put all the partition column expressions at the front of the list of fields, and then put the selected fields after that. But obviously things need to match up, so now we create a correct sized `Vec` of `None`, and then fill it in with the correct expressions in the correct positions, and then error if something didn't get filled in.

Reply:
I think that would work, but a couple of notes:
1. My code doesn't do any linear searching I don't think. Maybe you are thinking that `field_index` does a search, but it doesn't because it's just calling `get_index_of` on the `IndexMap` in the schema which is `O(1)` (see [here](https://docs.rs/indexmap/latest/indexmap/map/struct.IndexMap.html#method.get_index_of))
2. Your solution actually _does_ search `partition_columns` once for each field in the schema
3. We only need to invoke the expression evaluator if there are partition columns, so we wouldn't actually get rid of the `if partition_fields.is_empty()` call

So overall I'm not yet convinced this is better :)

Reply:
On the plus side it does reduce some of the overhead of building the read_schema and the partition columns. Let me think if we can get the best of both worlds.

Reply:
Okay, I implemented something a little better that what was there before, mostly just cleaning up the code above here to only make one pass over the total schema and separate things into the read schema or into partition columns.

In doing that I understood your suggestion more, which is that we'd _always_ build expressions to eval, but they'd just be all `Column` expressions if there where no partition columns, and then we'd always invoke the expression evaluator.

That would work, but it seems like a lot of work for what amounts to a no-op in the case we have no partition columns.

Reply:
A couple things to point out:
1. I believe the original code is incorrect in case the output schema doesn't include all partition columns. Your code doesn't make it worse, but my suggestion avoided the pitfall by only caring about partition columns the query actually mentions. If we want to validate the partition schema vs. the table schema, we could do that -- but I'm not sure it's worth the trouble?
2. My suggestion hoisted a lot of repeated work out of inner loops. We could probably adapt that optimization to your approach as well.

The above are orthogonal issues, tho. Thinking more, the essential difference between the two partition column handling approaches is: 
* Your approach is "out of place" -- it partitions the set of field indexes in the first pass, uses a second pass to "inject" each field into its correct location, and a third pass verifies no gaps remain. 
* My approach is "in place" -- the first pass leaves a mix of "resolved" and "unresolved" columns in the first pass, and a second pass "resolves" the partition columns. No gaps possible.

Both have the same big-O complexity -- `O(n*k)` where `n` is field count and `k` is field name length -- so we should probably just decide which is simpler to understand, code up, and maintain?

<details>

<summary>Detailed responses</summary>
 
> My code doesn't do any linear searching... it's just calling `get_index_of` on the `IndexMap` in the schema which is `O(1)`

You're right, I forgot we had added that optimization when I wrote that. And had rediscovered by the time I finished writing the overall suggestion... but didn't remember to go back and edit this part.

> We only need to invoke the expression evaluator if there are partition columns

Oh! I forgot about the early-out in case there are no partition columns. Easy enough to add back in. Keep in mind we'll eventually need to add rowid support as well, so whatever we do should be extensible beyond just partition columns.

> Your solution actually does search `partition_columns` once for each field in the schema

I don't think there's any escaping that, if we want to separate read schema columns from special columns? The original code had to do it as well. We _could_ make a map or set to "accelerate" the search, but there are usually 0-2 partition columns and ~never more than 4, so a simple linear search will almost certainly be faster in practice.

Assuming we have any special columns at all, there's _also_ no escaping to make at least one additional pass to reassemble the output schema after the read is complete. I believe the current code makes four passes in all (split, fill-None, inject, flatten) while mine makes two in all (left/right, flatten). Given two approaches with the same big-O, I would generally prefer the one with simpler code (hence my suggestion).

</details>

Reply:
> My code doesn't do any linear searching... it's just calling `get_index_of` on the `IndexMap` in the schema which is `O(1)`

One additional observation: The field lookups may be `O(1)` in the size of the map, but they're still a string search (`O(k)` in size of string). The first improvement I considered (before I settled on the in-place method) used index maps (powered by `iter().enumerate()`) to avoid that particular overhead:

<details>

```rust
let mut partition_fields: Vec<(usize, &String)> = !vec[];
let mut read_fields: Vec<(usize, StructField)> = !vec[];
self.schema.fields().enumerate().for_each(|(i, field)| {
  if partition_columns.contains(&field.name) {
    partition_fields.append((i, field.name()));
  } else {
    read_fields.append((i, field));
  }
})
```
... and then
```rust
for (index, field) in read_fields {
  fields[index] = ...;
}
for (index, name) in partition_fields {
  fields[index] = ...;
}
```  

</details>

However, then I realized that (a) the read column list will usually contain the majority of columns; and (b) the read column list could be populated _before_ we perform the read -- it's just a name-based projection! Thus, it seemed better to place the read columns directly in the final `fields` list, and that quickly evolved into the two-pass "in place" solution I ended up proposing.

Reply:
Okay, I've updated this now with something close to what you suggested. Changes from that include:
1. Using an `enum` rather than `Either`
2. only building and evaluating expressions if we actually have partition columns
3. Not building the `Column` expressions for "read" columns unless we have partition columns

I think this works pretty well. We build an unneeded vector of all `ColumnType::Selected` in the case that no partition column is selected, but that should be pretty cheap, and we have to iterate all the fields no matter what, so that's no extra work.